### PR TITLE
Fragment-count collection for ModelSegments pipeline.

### PR DESCRIPTION
**Pull request by samuelklee**
@asmirnov239 I made some additions to your code to add HDF5 writing and make TSV writing consistent.  I used an OverlapDetector instead of the CachedBinarySearchIntervalList and I think there was no performance hit.  I also went ahead and added HDF5Utils, which has some methods that will be used in later PRs.  

Let's consider your first commit reviewed by me; could you take a look at my changes in the second commit?

### CODE REVIEW

**Comment by mbabadi** [comment]
@samuelklee @asmirnov239 I still haven't had a chance to look at this PR and I've been pretty much out of your discussion regarding the new read count format -- in any event, is there a chance to report the average fragment-based GC content for each bin as well, i.e. (sum_j GC_content_of_fragment_j) / (total number of fragments)?

**Comment by samuelklee** [comment]
I think that the full version of the binned read-count collection that @asmirnov239 is working on could be easily modified to give you what you want.  Let's keep this tool as simple as possible for now.

However, something that would be much easier to change in this code (and might have a bigger effect) would be adding counts to all bins that overlap each fragment.  It would be interesting to see how this changes the statistics of the counts.  If we have some bandwidth, we can try experimenting with this before release.

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/coverage/readcount/SimpleCountCollection.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+/**
+ * Simple data structure to pass and read/write a List of {@link SimpleCount} objects.
+ *
+ * @author Samuel Lee &lt;slee@broadinstitute.org&gt;
+ */
+public final class SimpleCountCollection extends SampleLocatableCollection<SimpleCount> {
+    enum SimpleCountTableColumn {
```
Comment (asmirnov239) [inline src/main/java/org/broadinstitute/hellbender/tools/copynumber/coverage/readcount/SimpleCountCollection.java]: You can make this private

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/CollectFragmentCounts.java [hunk 2]
```diff
@@ -39,20 +39,25 @@
...
+        oneLineSummary = "Collect fragment counts.",
         programGroup = CopyNumberProgramGroup.class
 )
-public class CollectFragmentCounts extends ReadWalker {
+public final class CollectFragmentCounts extends ReadWalker {
+    private static final Logger logger = LogManager.getLogger(CollectFragmentCounts.class);
 
-    private static Logger logger = LogManager.getLogger(CollectFragmentCounts.class);
-    private SAMSequenceDictionary sequenceDictionary = getBestAvailableSequenceDictionary();
+    private static final int DEFAULT_MINIMUM_MAPPING_QUALITY = 30;
```
[the remark this answers is not in the document]
Comment (samuelklee) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/CollectFragmentCounts.java]: You might be right.  I simply set this to 30 to match the value used in CollectAllelicCounts.  Seems to be OK in preliminary evaluations, but this is the sort of thing that @katevoss can investigate in her parameter sweep.

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/CollectFragmentCounts.java [hunk 3]
```diff
@@ -96,85 +107,67 @@ public boolean requiresIntervals() {
...
-
-        final IndexRange intersectionRange = cachedIntervalList.findIntersectionRange(centerFragmentLocation);
-        if (intersectionRange.size() > 1) {
-            // should not reach here since intervals are checked for overlapping;
-            // doing a check to protect against future code changes
-            throw new GATKException.ShouldNeverReachHereException("At most one interval can intersect with a center of a fragment.");
+
+        try {
+            //getting the center of the fragment
+            //TODO make sure that center calculation always returns valid values
```
[the remark this answers is not in the document]
Comment (samuelklee) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/CollectFragmentCounts.java]: Simply that the center is within the contig.  I was getting some weird edge cases that returned negative fragment centers---hence the try-except, which absorbs these and emits a warning.  However, we could create a contig-length map and check every fragment center.  I'll add a bit to the comment to make things more clear.

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/coverage/readcount/HDF5SimpleCountCollection.java [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+import java.util.Arrays;
+import java.util.List;
+
+final class HDF5SimpleCountCollection implements SampleMetadata {
+    private static final String SAMPLE_NAME_PATH = "/sample_metadata/sample_name";
+    private static final String INTERVALS_GROUP_NAME = "/intervals";
+    private static final String COUNTS_PATH = "/counts/values";
+
+    private final HDF5File file;
+    private Lazy<List<SimpleInterval>> intervals;
```
[the remark this answers is not in the document]
Comment (samuelklee) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/coverage/readcount/HDF5SimpleCountCollection.java]: Done.

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/coverage/readcount/HDF5SimpleCountCollection.java [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+
+    List<SimpleInterval> getIntervals() {
+        return intervals.get();
+    }
+
+    /**
+     * @return single-row matrix containing the counts
+     */
+    RealMatrix getCounts() {
+        return new Array2DRowRealMatrix(file.readDoubleMatrix(COUNTS_PATH));
```
[the remark this answers is not in the document]
Comment (samuelklee) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/coverage/readcount/HDF5SimpleCountCollection.java]: Done, with sample name too.

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/coverage/readcount/SimpleCountCollection.java (hunk 1 above)
Comment (samuelklee) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/coverage/readcount/SimpleCountCollection.java]: These have been package-protected in the other SampleLocatableCollection classes, so I'll just keep it as is to be consistent.  At some point we may actually want to make all of these public, instead.

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/HDF5Utils.java [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+                        contigNames[(int) matrix[IntervalField.CONTIG_INDEX.index][i]],
+                        (int) matrix[IntervalField.START.index][i],
+                        (int) matrix[IntervalField.END.index][i])))
+                .collect(Collectors.toList());
+    }
+
+    public static <T extends SimpleInterval> void writeIntervals(final HDF5File file,
+                                                                 final String path,
+                                                                 final List<T> intervals) {
+        final String[] contigNames = intervals.stream().map(SimpleInterval::getContig).distinct().toArray(String[]::new);
```
[the remark this answers is not in the document]
Comment (samuelklee) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/HDF5Utils.java]: Good catch, I think I was just being lazy when I initially wrote this code.  How about this?

    final Map<String, Double> contigNamesToIndexMap = new LinkedHashMap<>();
        final double[][] matrix = new double[NUM_INTERVAL_FIELDS][intervals.size()];
        for (int i = 0; i < intervals.size(); i++) {
            final SimpleInterval interval = intervals.get(i);
            contigNamesToIndexMap.putIfAbsent(interval.getContig(), (double) contigNamesToIndexMap.keySet().size());
            matrix[IntervalField.CONTIG_INDEX.index][i] = contigNamesToIndexMap.get(interval.getContig());
            matrix[IntervalField.START.index][i] = interval.getStart();
            matrix[IntervalField.END.index][i] = interval.getEnd();
        }
        file.makeDoubleMatrix(path + INTERVAL_MATRIX_SUB_PATH, matrix);
        file.makeStringArray(path + INTERVAL_CONTIG_NAMES_SUB_PATH,
                contigNamesToIndexMap.keySet().toArray(new String[contigNamesToIndexMap.keySet().size()]));

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/HDF5Utils.java [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+        for (int i = 0; i < intervals.size(); i++) {
+            final SimpleInterval interval = intervals.get(i);
+            matrix[IntervalField.CONTIG_INDEX.index][i] = contigNamesToIndexMap.get(interval.getContig());
+            matrix[IntervalField.START.index][i] = interval.getStart();
+            matrix[IntervalField.END.index][i] = interval.getEnd();
+        }
+        file.makeDoubleMatrix(path + INTERVAL_MATRIX_SUB_PATH, matrix);
+    }
+
+    private static final String NUMBER_OF_ROWS_SUB_PATH = "/num_rows";
```
[the remark this answers is not in the document]
Comment (samuelklee) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/HDF5Utils.java]: Done, added some comments too.

File: src/test/java/org/broadinstitute/hellbender/tools/copynumber/CollectFragmentCountsIntegrationTest.java [hunk 8]
```diff
@@ -19,39 +23,108 @@
...
         return new Object[][] {
-                {NA12878_BAM, NA12878_FRAGMENT_COUNT_EXPECTED_OUTPUT}
+                {NA12878_BAM, NA12878_FRAGMENT_COUNTS_EXPECTED_OUTPUT}
         };
     }
 
+    @Test(expectedExceptions = IllegalArgumentException.class)
+    public void testIntervalSetRule() {
+        final File resultOutputFile = createTempFile("collect-fragment-counts-test", ".tsv");
+        final ArgumentsBuilder argsBuilder = new ArgumentsBuilder()
```
[the remark this answers is not in the document]
Comment (samuelklee) [reply src/test/java/org/broadinstitute/hellbender/tools/copynumber/CollectFragmentCountsIntegrationTest.java]: Eh, I think this is OK for test code.  A lot of our integration tests use the `final String[] arguments = {"-" + ARGUMENT_NAME, ARGUMENT_VALUE, ...}` pattern---this is just a nicer way.  But no need to extract arguments common to each test, as I think this actually makes things less readable.

File: src/test/java/org/broadinstitute/hellbender/tools/copynumber/CollectFragmentCountsIntegrationTest.java [hunk 9]
```diff
@@ -19,39 +23,108 @@
...
+                .addInput(inputBAMFile)
+                .addArgument("L", INTERVALS_FILE.getAbsolutePath())
+                .addArgument("imr", IntervalMergingRule.OVERLAPPING_ONLY.toString())
+                .addArgument(CollectFragmentCounts.OUTPUT_FORMAT_SHORT_NAME, CollectFragmentCounts.OutputFormat.HDF5.toString())
+                .addOutput(resultOutputFile);
+        runCommandLine(argsBuilder);
+        final SimpleCountCollection expectedCounts = SimpleCountCollection.read(expectedOutputFile);
+        final SimpleCountCollection resultCounts = SimpleCountCollection.read(new HDF5File(resultOutputFile));
+        Assert.assertEquals(expectedCounts, resultCounts);
+        Assert.assertNotSame(expectedCounts, resultCounts);
```
[the remark this answers is not in the document]
Comment (samuelklee) [reply src/test/java/org/broadinstitute/hellbender/tools/copynumber/CollectFragmentCountsIntegrationTest.java]: Hmm, I could've sworn I saw some of our other tests doing this (perhaps to protect against changes in the test code).  But now I can't find an example...I agree that it's overkill, will remove.

File: src/test/java/org/broadinstitute/hellbender/tools/copynumber/CollectFragmentCountsIntegrationTest.java [hunk 10]
```diff
@@ -61,10 +134,9 @@ public void test(final File inputBAMFile, final File expectedOutputFile) {
     }
 
     @Test(dataProvider = "artificialReadsData")
-    public void testFragmentCenterComputation(final GATKRead read, final int expectedFragmentCenter) {
-
-        final int actualFragmentCenter = CollectFragmentCounts.ReadOrientation.getCenterOfFragment(read);
-        Assert.assertEquals(actualFragmentCenter, expectedFragmentCenter);
+    public void testFragmentCenterComputation(final GATKRead read, final int expectedFragmentCenterPosition) {
+        final SimpleInterval expectedFragmentCenter = new SimpleInterval(read.getContig(), expectedFragmentCenterPosition, expectedFragmentCenterPosition);
+        final SimpleInterval resultFragmentCenter = CollectFragmentCounts.ReadOrientation.getFragmentCenter(read);
```
[the remark this answers is not in the document]
Comment (samuelklee) [reply src/test/java/org/broadinstitute/hellbender/tools/copynumber/CollectFragmentCountsIntegrationTest.java]: Good point, but a lot of our test code uses "result".  Maybe some boy scout could go through the CNV test code and make everything consistent with the testng convention...will leave for now.

File: src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/HDF5Utils.java [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+
+        IntervalField(final int index) {
+            this.index = index;
+        }
+    }
+    private static final int NUM_INTERVAL_FIELDS = IntervalField.values().length;
+
+    private HDF5Utils() {}
+
+    public static List<SimpleInterval> readIntervals(final HDF5File file,
```
[the remark this answers is not in the document]
Comment (samuelklee) [reply src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/HDF5Utils.java]: Added some javadoc.  I think many of the public methods in other classes are self-documenting, would rather not clutter the code too much.

**Comment by samuelklee** [comment]
Thanks, back to you @asmirnov239!

**Pull request merged**